cpp/include/raft/core/resource/user_resource.hpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <raft/core/resource/resource_types.hpp>
+#include <raft/core/resources.hpp>
+
+#include <memory>
+#include <typeindex>
+
+namespace raft::resource {
+
+class user_resource : public resource {
+ public:
+  user_resource()                    = default;
+  ~user_resource() noexcept override = default;
+  auto get_resource() -> void* override { return this; }
+
+  template <typename ResourceT>
+  auto load() -> ResourceT*
+  {
+    std::lock_guard<std::mutex> _(lock_);
+    auto key = std::type_index{typeid(ResourceT)};
+    auto pos = map_.find(key);
+    if (pos != map_.end()) { return reinterpret_cast<ResourceT*>(pos->second.get()); }
+    auto store_ptr = new ResourceT{};
+    map_[key] =
+      std::shared_ptr<void>(store_ptr, [](void* ptr) { delete reinterpret_cast<ResourceT*>(ptr); });
+    return store_ptr;
+  }
+
+ private:
+  std::unordered_map<std::type_index, std::shared_ptr<void>> map_{};
```

Comment:
While I was thinking about overheads of copying the resources handle in the gemm/matmul, it occurred to me that we hadn't discussed what should happen to user resources when the resources handle is copied. I see three sensible options:
  1. [current implementation] The whole map is shared between the old and the new handle. Low overhead, but setting a user resource in one handle affects the other handle as well.
  2. Do a shallow copy of the map when copying the resources handle. Incurs more unwanted overheads, but the behavior of individual user resources is the same as of the  compile-time-registered resources (setting an individual resource affects only the current handle).
  3. Drop the user resource map on copy. Incurs a low overhead during copy. Resets the state of the user resources, not sure for better or worse (e.g. the kernel cache resources  are independent).

@cjnolet, do you have a preference here?

Reply:
@achirkin maybe we could provide a flag for this? Perhaps even have a predefined element in the map of user resources containing this flag? This way they user could set/change the flag using raft::resource::set_user_resource_copy_mode()` or something? We can default it to the least overhead option.

Reply:
@cjnolet do you mean a runtime flag? I've been thinking about that. At first glance, it seems nice to have this flexibility. However, I think, this has a potential to do more harm than good.
From what I've seen so far, most of the "users" of the user_resource are within raft itself (e.g. caching in the algorithms). When we add these use cases, we keep in mind particular behavior of the resource (e.g. cached kernel parameters are shared across the copied handles or kept private). I can imagine this behavior affecting the performance of the algorithms. If an external user is able to change this behavior at runtime (because they want a particular behavior in their piece of code) it will have an adverse effect on overall performance of the other parts of raft codebase.
Therefore, I think it's reasonable to decide on this behavior once and for all and design the uses cases with the chosen behavior in mind.

Reply:
> From what I've seen so far, most of the "users" of the user_resource are within raft itself (e.g. caching in the algorithms).

Your comment here makes me struggle with the name a little bit, because I was absolutely thinking we would want to invite users to use a "user resource". One of the reasons we didn't use an unordered map for the other resources is because lookup has shown to be slow, especially when the map is small. 

Rather than a "user resource", this seems to be more of a "runtime shared resource cache". I'd prefer to name it appropriately based on its use and I think that name is generalized enough to tell a potential user (internal or external) what it's for. Docs can help describe it even better for anyone who might have further confusion on the name.

Reply:
I think it's ok to invite users to use this resource, but you're right the name implies it's made specifically for users, which is wrong. How about "custom_resource"?

To me, the only difference of the custom_resource to the other resources is that raft can't allocate a slot for it at compile time. I've tried to reflect it in the updated name and docs. How does it look to you now?

Also regarding the resource copy policy: I suggest we stick to the status quo now (option 1) and revise it when we decide something on `raft::resources` and thread-safety issue.